mypy/typeops.py

```diff
@@ -312,46 +312,40 @@ class B(A): pass
             original_type = erase_to_bound(self_param_type)
         original_type = get_proper_type(original_type)
 
-        all_ids = func.type_var_ids()
+        # Find which of method type variables appear in the type of "self".
+        self_ids = {tv.id for tv in get_all_type_vars(self_param_type)}
+        self_vars = [tv for tv in func.variables if tv.id in self_ids]
+
+        # Solve for these type arguments using the actual class or instance type.
         typeargs = infer_type_arguments(
-            func.variables, self_param_type, original_type, is_supertype=True
+            self_vars, self_param_type, original_type, is_supertype=True
         )
         if (
             is_classmethod
-            # TODO: why do we need the extra guards here?
             and any(isinstance(get_proper_type(t), UninhabitedType) for t in typeargs)
             and isinstance(original_type, (Instance, TypeVarType, TupleType))
         ):
-            # In case we call a classmethod through an instance x, fallback to type(x)
+            # In case we call a classmethod through an instance x, fallback to type(x).
             typeargs = infer_type_arguments(
-                func.variables, self_param_type, TypeType(original_type), is_supertype=True
+                self_vars, self_param_type, TypeType(original_type), is_supertype=True
             )
 
-        ids = [tid for tid in all_ids if any(tid == t.id for t in get_type_vars(self_param_type))]
-
-        # Technically, some constrains might be unsolvable, make them <nothing>.
+        # Update the method signature with the solutions found.
+        # Technically, some constraints might be unsolvable, make them <nothing>.
         to_apply = [t if t is not None else UninhabitedType() for t in typeargs]
-
-        def expand(target: Type) -> Type:
-            return expand_type(target, {id: to_apply[all_ids.index(id)] for id in ids})
-
-        arg_types = [expand(x) for x in func.arg_types[1:]]
-        ret_type = expand(func.ret_type)
-        variables = [v for v in func.variables if v.id not in ids]
+        func = expand_type(func, {tv.id: arg for tv, arg in zip(self_vars, to_apply)})
+        variables = [v for v in func.variables if v not in self_vars]
```

Comment:
Seeing this whole "expand a type with certain type vars then remove those certain type vars" dance feels weird. Is there a reason `expand_type` doesn't remove the replaced type variables?

Reply:
OK, it is time for type theory lesson. There are two kinds of things that people usually call "generic types": type constructors, and polymorphic types. This terminology is still vague (and I am not sure it is standard), so it is better to explain on examples. For example, `Callable` is a type constructor, by itself it is not a valid type, it expects some type arguments, so e.g. `Callable[[int], int]` is a valid type. Now when you write:
```python
def foo(x: T) -> T: ...
```
assuming `T` is not bound in current scope, this defines a polymorphic type for `foo`. Such type is valid by itself and does not expect any type arguments. You can imagine it as an infinite intersection or an infinite overload `forall T . Callable[[T], T]`. 

(An implementation note: In an ideal world `variables` should not be a property of `CallableType`, we would have `ForAllType(variables, template)`, and there would be no `fallback` in callable types, in an ideal world type of `foo` would be `ForAllType([T], IntersectionType(Callable[[T], T], Instance("types.FunctionType")))`, but we are not living in an ideal world.)

If we are dealing with type constructor (in particular with generic type aliases), we never need to "remove certain type vars" (what would it even mean). The only thing we should do is to formally substitute type parameters expected by constructor with type arguments we got. For example, with `A[T] = Callable[[T], T]` (it is sad that it is not a valid syntax, having explicit vars for type aliases would save a lot of ambiguity/confusion), and with `A[int]` we just do `expand_type(Callable[[T], T], {T -> int}) = Callable[[int], int]`, end of story.

Now when we talk about type inference this usually involves several steps:
1. Given an infinite overload (ranging over some type variables), figure out how types of arguments passed in `CallExpr` narrow down range of each type variable (this is called inferring constraints, done in `constraints.py`). For `foo(42)` this would be `T :> int`.
2. Find the matching overload by solving those constraints (done in `solve.py`). In the above example `T = int` is best solution.
3. Construct the type of matching overload. This is usually done in `applytype.py` and has three sub-steps:
    a. Validate that the solution found actually satisfies type variable bounds, values, etc.
    b. Call `expand_type()` on the r.h.s. of `forall` expression with solutions found.
    c. Update free variables on the l.h.s of `forall` expression (usually this means just remove some, but not always).

So what you are seeing here is some ad-hoc version of this logic. Or course it all can be refactored to reduce code duplication but FWIW I don't think it is a big deal in this case, since the amount of code duplication is really small here.

Reply:
And one more clarifying comments on why we don't have the validation step (3a) in the case of type constructor: this is because validation can be complete during semantic analysis, see `semanal_typeargs.py`.